### PR TITLE
Various fixes and enhancements

### PR DESCRIPTION
**Pull request by Shadowghost**
### Bug Fixes
  - Fix unauthorized users not being skipped when syncing
  - Fix MediaType metadata
  - Fix 4:3 resolution handling
  - Don't fail tasks if one user fails

  ### Features
  - Implement HDR metadata
  - Implement metadata for episodes
  - Implement 480p fallback (minimum for metadata on trakt)
  - Switch to GUIDs (this requires reauthentication of all users)

### CODE REVIEW

**Review by h1dden-da3m0n — DISMISSED** [review]
Overall the changes LGTM, but I am not too familiar with this plugin nor Trak.tv, so cannot comment on the functional aspects of the changes

File: Trakt/Api/TraktApi.cs [hunk 1]
```diff
@@ -279,6 +279,8 @@ public async Task<List<TraktScrobbleResponse>> SendEpisodeStatusUpdateAsync(Epis
                 traktMovieCollected.AudioChannels = audioStream.GetAudioChannels();
                 traktMovieCollected.Audio = audioStream.GetCodecRepresetation();
                 traktMovieCollected.Resolution = m.GetDefaultVideoStream().GetResolution();
+                traktMovieCollected.Is3D = m.Is3D;
+                traktMovieCollected.Hdr = m.GetDefaultVideoStream().GetHdr();
```
Comment (cvium) [inline Trakt/Api/TraktApi.cs:283]: Calling `GetDefaultVideoStream` twice

File: Trakt/Api/TraktApi.cs [hunk 2]
```diff
@@ -350,14 +352,16 @@ public async Task<List<TraktScrobbleResponse>> SendEpisodeStatusUpdateAsync(Epis
             var audioStream = episode.GetMediaStreams().FirstOrDefault(stream => stream.Type == MediaStreamType.Audio);
 
             if (useProviderIds && HasAnyProviderTvIds(episode) &&
-                (!episode.IndexNumber.HasValue || !episode.IndexNumberEnd.HasValue ||
-                 episode.IndexNumberEnd <= episode.IndexNumber))
+                (!episode.IndexNumber.HasValue
+                    || !episode.IndexNumberEnd.HasValue
```
Comment (cvium) [inline Trakt/Api/TraktApi.cs:356]: indentation is off

File: Trakt/Api/TraktController.cs [hunk 3]
```diff
@@ -62,11 +62,13 @@ public async Task<ActionResult<object>> TraktDeviceAuthorization([FromRoute] str
         _logger.LogInformation("TraktDeviceAuthorization request received");
 
         // Create a user if we don't have one yet - TODO there should be an endpoint for this that creates a default user
-        var traktUser = UserHelper.GetTraktUser(userId);
+        var userGuid = new Guid(userId);
```
Comment (cvium) [inline Trakt/Api/TraktController.cs:65]: I think `Guid.Parse` better describes what happens, but idc if you change it

File: Trakt/Extensions.cs [hunk 4]
```diff
@@ -121,7 +121,38 @@ public static bool MetadataIsDifferent(this TraktMovieCollected collectedMovie,
...
+
+    /// <summary>
+    /// Checks if metadata of new collected movie is different from the already collected.
+    /// </summary>
+    /// <param name="collectedEpisode">The <see cref="TraktEpisodeCollected"/>.</param>
+    /// <param name="episode">The <see cref="Movie"/>.</param>
+    /// <returns><see cref="bool"/> indicating if the new episode has different metadata to the already collected.</returns>
+    public static bool MetadataIsDifferent(this TraktEpisodeCollected collectedEpisode, Episode episode)
+    {
+        var audioStream = episode.GetMediaStreams().FirstOrDefault(x => x.Type == MediaStreamType.Audio);
```
Comment (cvium) [inline Trakt/Extensions.cs:135]: `GetMediaStreams` also contains the default video stream, but you are calling them thrice (combined). There's no caching afaict.

File: Trakt/Extensions.cs [hunk 5]
```diff
@@ -121,7 +123,40 @@ public static bool MetadataIsDifferent(this TraktMovieCollected collectedMovie,
...
+    /// <param name="collectedEpisode">The <see cref="TraktEpisodeCollected"/>.</param>
+    /// <param name="episode">The <see cref="Episode"/>.</param>
+    /// <returns><see cref="bool"/> indicating if the new episode has different metadata to the already collected.</returns>
+    public static bool MetadataIsDifferent(this TraktEpisodeCollected collectedEpisode, Episode episode)
+    {
+        var mediaStreams = episode.GetMediaStreams();
+        var defaultVideoStream = mediaStreams.FirstOrDefault(x => x.Index == episode.DefaultVideoStreamIndex);
+        var audioStream = mediaStreams.FirstOrDefault(x => x.Type == MediaStreamType.Audio);
+
+        var resolution = defaultVideoStream.GetResolution();
```
Comment (cvium) [inline Trakt/Extensions.cs:141]: This will crash if defaultvideostream is null. If it cannot be null, use `First()`

File: Trakt/Extensions.cs [hunk 6]
```diff
@@ -146,29 +181,23 @@ public static bool MetadataIsDifferent(this TraktMovieCollected collectedMovie,
...
         }
 
-        if (videoStream.Width.Value >= 1900)
+        if (videoStream.Width.Value >= 1400)
         {
             return videoStream.IsInterlaced ? TraktResolution.hd_1080i : TraktResolution.hd_1080p;
         }
 
-        if (videoStream.Width.Value >= 1270)
+        if (videoStream.Width.Value >= 950)
```
Comment (cvium) [inline Trakt/Extensions.cs:189]: looks very arbitrary?

Comment (Shadowghost) [reply Trakt/Extensions.cs:189]: 720p 4:4 is 960×720 and I gave it 10px leeway

Comment (Shadowghost) [reply Trakt/Extensions.cs]: Moved them to consts and added some remarks to why they are as they are.

File: Trakt/Api/TraktController.cs [hunk 7]
```diff
@@ -62,11 +62,13 @@ public async Task<ActionResult<object>> TraktDeviceAuthorization([FromRoute] str
         _logger.LogInformation("TraktDeviceAuthorization request received");
 
         // Create a user if we don't have one yet - TODO there should be an endpoint for this that creates a default user
-        var traktUser = UserHelper.GetTraktUser(userId);
+        var userGuid = Guid.Parse(userId);
```
Comment (crobibero) [inline Trakt/Api/TraktController.cs:65]: I would change all of the routes to accept `Guid` instead of `string`

File: build.yaml [hunk 8]
```diff
@@ -15,10 +15,11 @@ artifacts:
...
 
-  ### Dependency updates ###
-  - chore(deps): bump StyleCop.Analyzers from 1.2.0-beta.406 to 1.2.0-beta.435 (#139) @dependabot
-
-  ### CI & build changes ###
-  - fix: meta ci workflows (#145) @h1dden-da3m0n
+  ### Features ###
+  - Implement HDR metadata (#154) @Shadowghost
+    - Implement metadata for episodes (#154) @Shadowghost
+  - Switch to GUIDs (this requires reauthentication of all users) (#154) @Shadowghost
```
Comment (crobibero) [inline build.yaml:25]: This is managed by the release-draft PR so it'll be fully overwritten

Comment (Shadowghost) [reply build.yaml:25]: Yeah, we already discussed this on matrix

File: Trakt/Api/TraktController.cs [hunk 9]
```diff
@@ -51,22 +51,23 @@ public class TraktController : ControllerBase
...
-    /// <param name="userId">The user id of the user connecting to trakt.tv.</param>
+    /// <param name="userGuid">The GUID of the user connecting to trakt.tv.</param>
     /// <response code="200">Authorization code requested successfully.</response>
     /// <returns>The trakt.tv authorization code.</returns>
-    [HttpPost("Users/{userId}/Authorize")]
+    [HttpPost("Users/{userGuid}/Authorize")]
     [Authorize(Policy = "RequiresElevation")]
     [ProducesResponseType(StatusCodes.Status200OK)]
-    public async Task<ActionResult<object>> TraktDeviceAuthorization([FromRoute] string userId)
+    public async Task<ActionResult<object>> TraktDeviceAuthorization([FromRoute] Guid userGuid)
```
Comment (crobibero) [inline Trakt/Api/TraktController.cs:60]: Just a note- you could have kept the parameter as `userId`

**Review by crobibero — APPROVED** [review]

**Pull request merged**